loopy/target/c/codegen/expression.py

```diff
@@ -516,13 +508,25 @@ def map_power(self, expr, type_context):
             clbl = self.codegen_state.ast_builder.known_callables["pow"]
             clbl = clbl.with_types({0: tgt_dtype, 1: exponent_dtype},
                     self.codegen_state.callables_table)[0]
+
             self.codegen_state.seen_functions.add(
                     SeenFunction(
                         clbl.name, clbl.name_in_target,
                         (base_dtype, exponent_dtype),
                         (tgt_dtype,)))
-            return var(clbl.name_in_target)(self.rec(expr.base, type_context),
-                    self.rec(expr.exponent, type_context))
+
+            common_dtype = np.find_common_type(
+                [], [dtype.numpy_dtype for id, dtype in clbl.arg_id_to_dtype.items()
+                     if (id >= 0 and dtype is not None)])
+            from loopy.types import NumpyType
+            dtype = NumpyType(common_dtype)
+            inner_type_context = dtype_to_type_context(
+                    self.kernel.target, dtype)
+
+            return var(clbl.name_in_target)(
+                self.rec(expr.base, inner_type_context, dtype),
+                self.rec(expr.exponent, inner_type_context, dtype)
+            )
```

Comment:
I feel like this code is duplicating logic and could be replaced with

```python
return self.rec(
    var("pow")(expr.base, expr.exponent),
    type_context
)
```

but I get key errors for "pow" in `self.codegen_state.callables_table`. (The same holds if I keep the `seen_function` business.) Not sure what the deal is here; I'd appreciate any guidance!

Reply:
Have you tried your version with `clbl.name_in_target`? I don't think `pow` has a chance of getting turned into a `ResolvedFunction` at this stage:
https://github.com/inducer/loopy/blob/ef7c4119b672f5e4e09632a80f7548e80dd9e126/loopy/symbolic.py#L817-L836
The way I see it, the `known_callables["pow"]` hacks around that.

IMO, recursing on the resolved callable might work, but I'm also not hating the current version.

Maybe @kaushikcfd can add some perspective, too.

Reply:
> I don't think pow has a chance of getting turned into a ResolvedFunction at this stage:

Yep that's my assessment too. Which is the reason we had the `known_callables["pow"]` lying around.

Sorry, maybe I'm missing the context here, but why do we need the type casts? Isn't C's type-promotion along with the `name_in_target` enough?

Reply:
First, let me link to [my comment here](https://github.com/inducer/loopy/pull/438#issuecomment-873102901) (for some reason, I couldn't reply to this thread at the time).

> why do we need the type casts?

The multi-arg math functions in CL refuse to promote in the case of mixed-type inputs, raising compilation errors about ambiguous input types.

Reply:
Aah I see missed that (GH was not allowing to me reply to this as well. waiting for a couple minutes resolved it, strange), thanks for explaining! I'm fine with this approach too, but any reason why the way we had handled for `powf(32|64)` wasn't chosen for the other multi-arg callables. Is it because we would end up having too many `if-elses` leading to unmaintainable code?

Reply:
In my view (just from working on this PR), those premables for `pow` were a bit redundant since `with_types` already (nominally) implements the logic for type promotion---and it's more generic, e.g., to half precision types, etc. I imagine `powf32/powf64` arose for historical reasons, since (unlike the other math builtins) power is its own symbolic primitive.

Reply:
> and it's more generic, e.g., to half precision types, etc

Agreed that this is more general. I would have loved it if there weren't `(float) (0.0f)`, but I guess we could improve on it later. 

> powf32/powf64 arose for historical reasons

Those had arose just to solve the same issue: https://github.com/inducer/loopy/pull/186/commits/7e37f5effc01b4e685de2b0475600e18e3729049. But not quite as neatly .

Reply:
>  I would have loved it if there weren't `(float) (0.0f)`

Me as well, but at the moment the decision to cast occurs before recursing on the constant itself (at which point a `0` could be replaced with `0.0f`).